scripts/cleanup-azure-resources.sh

```
@@ -594,6 +667,93 @@ main() {
+            found_any=true
+            delete_resources "$resources_json"
+        fi
+
+        # Find Azure AD Applications and Service Principals with matching tags.
+        # AD app tags are string arrays with "key:value" format (not ARM key=value).
+        # Microsoft Graph doesn't support server-side tag filtering via OData, so we use
+        # the capi-test-run-id tag value (which equals CS_CLUSTER_NAME) as a displayName
+        # prefix filter (fast, server-side) and then verify tags client-side.
+        local tag_key="${TAG_FILTER%%=*}"
+        local tag_value="${TAG_FILTER#*=}"
+        local ad_tag_string="${tag_key}:${tag_value}"
```

Shellcheck (0.11.0)

[warning] ad_tag_string appears unused. Verify use (or export if used externally).

(SC2034)